File: tests/cs/spectral-norm.cs
using System;
using System.Diagnostics;
using System.Timers;
using System.Collections.Generic;

class HelloWorld {

    delegate void MyBlock();

    static void profile ( int count, string category, MyBlock f ) {
        long n = 1;
        double total = 0;
        while (true) {
            Stopwatch stopwatch = Stopwatch.StartNew();
            for ( long i = 0; i < n; i++ ) {
                f();
            }
            stopwatch.Stop();
            total = stopwatch.Elapsed.TotalSeconds;
            if (total >= 0.5 || n >= 1000000000L) break;
            double per = Math.Max(total / n, 1e-9);
            long next = (long)(0.5 / per * 1.2);
            next = Math.Min(next, 100L * n);
            next = Math.Max(next, n + 1);
            n = Math.Min(next, 1000000000L);
        }
        Console.WriteLine($"\"{category}\", {total / n}, {n}");
    }

    static int A(int i, int j) {
        return ((i + j) * (i + j + 1) / 2 + i + 1);
    }

    static double dot(double [] v, double [] u, int n) {
        int i;
        double sum = 0;
        for (i = 0; i < n; i++)
            sum += v[i] * u[i];
        return sum;
    }

    static void mult_Av(double [] v, double [] output, int n) {
        int i, j;
        double sum;
        for (i = 0; i < n; i++) {
            for (sum = j = 0; j < n; j++)
                sum += v[j] / A(i, j);
            output[i] = sum;
        }
    }

    static void mult_Atv(double [] v, double [] output, int n) {
        int i, j;
        double sum;
        for (i = 0; i < n; i++) {
            for (sum = j = 0; j < n; j++)
                sum += v[j] / A(j, i);
            output[i] = sum;
        }
    }

    static double [] tmp;
    static void mult_AtAv(double [] v, double [] output, int n) {
        mult_Av(v, tmp, n);
        mult_Atv(tmp, output, n);
    }

    static double testSnorm(int n) {
        double[] u = new double[n];
        double[] v = new double[n];
        tmp = new double[n];
        int i;
        for (i = 0; i < n; i++)
            u[i] = 1;
        for (i = 0; i < 2; i++) {
            mult_AtAv(u, v, n);
            mult_AtAv(v, u, n);
        }
        double result = Math.Sqrt(dot(u, v, n) / dot(v, v, n));
        tmp = null;
        u = null;
        v = null;
        return result;
    }


    static void Main() {
        profile(10, "spectral norm", () => {
            testSnorm(500);
        });
    }
}

File: tests/js/spectral-norm.js
function A(i,j) {
return 1/((i+j)*(i+j+1)/2+i+1);
}

function Au(u,v) {
for (var i=0; i<u.length; ++i) {
    var t = 0;
    for (var j=0; j<u.length; ++j)
    t += A(i,j) * u[j];
    v[i] = t;
}
}

function Atu(u,v) {
for (var i=0; i<u.length; ++i) {
    var t = 0;
    for (var j=0; j<u.length; ++j)
    t += A(j,i) * u[j];
    v[i] = t;
}
}

function AtAu(u,v,w) {
Au(u,w);
Atu(w,v);
}

function spectralnorm(n) {
var i, u=[], v=[], w=[], vv=0, vBv=0;
for (i=0; i<n; ++i) {
    u[i] = 1; v[i] = w[i] = 0;
}
for (i=0; i<2; ++i) {
    AtAu(u,v,w);
    AtAu(v,u,w);
}
for (i=0; i<n; ++i) {
    vBv += u[i]*v[i];
    vv  += v[i]*v[i];
}
return Math.sqrt(vBv/vv);
}

function timeStamp() {
	return performance.now();
}

function profile(tname, cnt, testFn) {
    let n = 1
    let total = 0
    while (true) {
        const start = timeStamp()
        for (let i = 0; i < n; ++i) {
            testFn()
        }
        total = timeStamp() - start
        if (total >= 500.0 || n >= 1000000000) break
        const per = Math.max(total / n, 1e-6)
        let next = Math.floor(500.0 / per * 1.2)
        next = Math.min(next, 100 * n)
        next = Math.max(next, n + 1)
        n = Math.min(next, 1000000000)
    }
    print('"' + tname + '", ' + (total / 1000.0 / n) + ', ' + n)
}

function performance_tests() {
	{
		profile("spectral norm",10,function(){
			spectralnorm(500);
		});
	}
	timeStamp();
}

performance_tests();